core/src/replay_stage.rs

```diff
@@ -3236,45 +3238,85 @@ impl ReplayStage {
+                        // 2. Our last vote was on a non-duplicate/confirmed slot
+                        // 3. Our last vote is now outside slot hashes history of the tip of fork
+                        // So, there was no hope of this last vote ever landing again.
+
+                        // In this case, we do want to obey threshold, yet try to register our vote on
+                        // the current fork, so we choose to vote at the tip of current fork instead.
+                        // This will not cause longer lockout because lockout doesn't double after 512
+                        // slots, it might be enough to get majority vote.
+                        reset_bank.map(|b| (b, SwitchForkDecision::SameFork))
+                    } else {
+                        // If we can't switch and our last vote was on a non-duplicate/confirmed slot, then
+                        // reset to the the next votable bank on the same fork as our last vote,
+                        // but don't vote.
+
+                        // We don't just reset to the heaviest fork when switch threshold fails because
+                        // a situation like this can occur:
+
+                        /* Figure 1:
+                                    slot 0
+                                        |
+                                    slot 1
+                                    /        \
+                        slot 2 (last vote)     |
+                                    |      slot 8 (10%)
+                            slot 4 (9%)
+                        */
+
+                        // Imagine 90% of validators voted on slot 4, but only 9% landed. If everybody that fails
+                        // the switch threshold abandons slot 4 to build on slot 8 (because it's *currently* heavier),
+                        // then there will be no blocks to include the votes for slot 4, and the network halts
+                        // because 90% of validators can't vote
+                        info!(
+                            "Waiting to switch vote to {},
+                            resetting to slot {:?} for now,
+                            switch proof stake: {},
+                            threshold stake: {},
+                            total stake: {}",
+                            heaviest_bank.slot(),
+                            reset_bank.as_ref().map(|b| b.slot()),
+                            switch_proof_stake,
+                            total_stake as f64 * SWITCH_FORK_THRESHOLD,
+                            total_stake
+                        );
+                        failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
+                            heaviest_bank.slot(),
+                            switch_proof_stake,
+                            total_stake,
+                        ));
+                        reset_bank.map(|b| (b, switch_fork_decision))
```

Comment:
nit: might be good to extract this into a separate `select_forks_failed_switch_threshold()`, because this logic is a bit long

Reply:
Done.